google/cloud/bigtable/internal/async_bulk_apply.cc

```diff
@@ -78,17 +71,33 @@ void AsyncRetryBulkApply::StartIterationIfNeeded(CompletionQueue cq) {
         self->OnRead(std::move(r));
         return make_ready_future(true);
       },
-      [self, cq](Status s) { self->OnFinish(cq, std::move(s)); });
+      [self, cq](Status const& s) { self->OnFinish(cq, s); });
 }
 
 void AsyncRetryBulkApply::OnRead(
     google::bigtable::v2::MutateRowsResponse response) {
   state_.OnRead(response);
 }
 
-void AsyncRetryBulkApply::OnFinish(CompletionQueue cq, Status status) {
-  state_.OnFinish(std::move(status));
-  StartIterationIfNeeded(std::move(cq));
+void AsyncRetryBulkApply::OnFinish(CompletionQueue cq, Status const& status) {
+  state_.OnFinish(status);
+  if (!state_.HasPendingMutations() ||
+      (!status.ok() && !rpc_retry_policy_->OnFailure(status))) {
+    promise_.set_value(std::move(state_).OnRetryDone());
+    return;
+  }
+
+  auto self = this->shared_from_this();
+  cq.MakeRelativeTimer(rpc_backoff_policy_->OnCompletion(status))
+      .then(
+          [self,
+           cq](future<StatusOr<std::chrono::system_clock::time_point>> result) {
+            if (auto tp = result.get()) {
+              self->StartIteration(std::move(cq));
+            } else {
+              self->promise_.set_value(std::move(self->state_).OnRetryDone());
```

Comment:
The double `self->` here is odd, consider (but this is your call) a private `OnRetryDone()` function